GH-103484: Fix broken links reported by linkcheck

rffontenelle opened the pull request:
This is another patch required to fix the current state of `make linkcheck` in Python Docs, see #103484.

This pull request fixes some broken links reported by `make linkcheck`. The backport to 3.11 has a few differences and I have already a patch ready for it, just waiting for any change in this one.

Find below the reported error and what solution I applied in this PR:

```
distributing/index.rst:128: [broken]  https://packaging.python.org/tutorials/packaging-projects/#creating-the-package-files: 
distributing/index.rst:127: [broken]  https://packaging.python.org/tutorials/packaging-projects/#packaging-python-projects:
distributing/index.rst:129: [broken]  https://packaging.python.org/tutorials/packaging-projects/#uploading-the-distribution-archives:
distributing/index.rst:130: [broken]  https://packaging.python.org/specifications/pypirc/:
```

The link is fine, but for some reason a newline in the doc resulted in being considered as broken for linkcheck, even though it is not broken in the documentation. I removed that newline and this made linkcheck happy.

```
library/stdtypes.rst:1607: [broken] http://www.unicode.org/versions/Unicode15.0.0/ch03.pdf#G53253: 'utf-8' codec can't decode byte 0xe2 in position 10: invalid continuation byte
library/stdtypes.rst:1767: [broken] https://www.unicode.org/versions/Unicode15.0.0/ch04.pdf#G91002: 'utf-8' codec can't decode byte 0xe2 in position 10: invalid continuation byte
library/stdtypes.rst:1906: [broken] https://www.unicode.org/versions/Unicode15.0.0/ch03.pdf#G34078: 'utf-8' codec can't decode byte 0xe2 in position 10: invalid continuation byte
```

That's sphinx-doc/sphinx#11041. I removed the anchor and added the section name next to it's [section] number so the reader has no doubt of what section the text is talking about.

```
whatsnew/changelog.rst:18176: [broken] https://: Invalid URL 'https://': No host supplied
```

This is the code sample `urllib.request.urlopen('https://...').` at Misc/NEWS.d/3.9.0a1.rst. Added it to ignored list as `'https:\/\/$'` (`$` to not match any other link).

```
howto/urllib2.rst:457: [broken] http://www.voidspace.org.uk/python/articles/authentication.shtml: HTTPConnectionPool(host='www.voidspace.org.uk', port=80): Max retries exceeded with url: /python/articles/authentication.shtml (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f1bb795e250>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution'))
```

www.voidspace.org.uk is down, so I replaced it with a Wayback Machine link. There was a code sample that also used this broken link and I replaced with a valid link: http://www.python.org following the previous example in the same file

```
whatsnew/2.6.rst:174: [broken] http://www.upfrontsoftware.co.za: HTTPSConnectionPool(host='www.upfrontsoftware.co.za', port=443): Max retries exceeded with url: / (Caused by SSLError(CertificateError("hostname 'www.upfrontsoftware.co.za' doesn't match either of 'agibase.com', 'gazette.co.za', 'icinga.siyavula.com', 'icinga.upfronthosting.co.za', 'lists.agibase.com', 'test.agibase.com', 'upfrontsoftware.co.za', 'www.agibase.com', 'www.gazette.co.za'")))
```

Fix removing 'www', replacing with https://upfrontsoftware.co.za

```
using/mac.rst:20: [broken] https://developer.apple.com/documentation/macos-release-notes/macos-12_3-release-notes#Python: Anchor 'Python' not found
bugs.rst:39: [broken] https://devguide.python.org/docquality/#helping-with-documentation: Anchor 'helping-with-documentation' not found
whatsnew/3.4.rst:1962: [broken] https://devguide.python.org/coverage/#measuring-coverage-of-c-code-with-gcov-and-lcov: Anchor 'measuring-coverage-of-c-code-with-gcov-and-lcov' not found
bugs.rst:41: [broken] https://devguide.python.org/documenting/#translating: Anchor 'translating' not found
library/gc.rst:101: [broken] https://devguide.python.org/garbage_collector/#collecting-the-oldest-generation: Anchor 'collecting-the-oldest-generation' not found
using/unix.rst:69: [broken] https://devguide.python.org/setup/#get-the-source-code: Anchor 'get-the-source-code' not found
```

These links lead to the expected anchors without issue, so I added ignore entries to these links.

```
whatsnew/3.8.rst:77: [broken] https://en.wikipedia.org/wiki/Walrus#/media/File:Pacific_Walrus_-_Bull_(8247646168).jpg: Anchor '/media/File:Pacific_Walrus_-_Bull_(8247646168).jpg' not found
```

The link is works, but the #/... is considered by linkcheck as invalid anchor. Added it to ignored anchors.

```
whatsnew/changelog.rst:16408: [broken] https://fishshell.com/docs/current/commands.html#source: Anchor 'source' not found
```

The `source` command is now on another page, so I updated the URL.

```
whatsnew/3.11.rst:1320: [broken] https://github.com/faster-cpython/ideas#published-results: Anchor 'published-results' not found
```

Anchors from Markdown files in GitHub repositories are not recognized, even though they work just fine. Hence I added this case to ignored links.

```
whatsnew/changelog.rst:15577: [broken] https://importlib-metadata.readthedocs.io/en/latest/changelog%20(links).html#v1-5-0: 404 Client Error: Not Found for url: https://importlib-metadata.readthedocs.io/en/latest/changelog%20(links).html
```

Updated the URL with the new page containing the versions history.

```
howto/functional.rst:1210: [broken] https://mitpress.mit.edu/sicp/: 404 Client Error: Not Found for url: https://mitpress.mit.edu/sicp/
```

Removing the trailing '/' solves the 404 Client Error.

However, there is another issue: The book is no longer freely available <s>(wayback machine disagrees)</s>, so I updated the text to say "The book can be found at" instead of "Full text at".

```
whatsnew/2.7.rst:2105: [broken] https://sourceware.org/gdb/current/onlinedocs/gdb/Python.html: 404 Client Error: Not Found for url: https://sourceware.org/gdb/current/onlinedocs/gdb/Python.html
```

Used Wayback Machine because the paragraph mentions GDB 7, so I linked to the latest GDB online docs available, from 2011.

```
using/windows.rst:554: [broken] https://support.enthought.com/hc/en-us/articles/360038600051-Canopy-GUI-end-of-life-transition-to-the-Enthought-Deployment-Manager-EDM-and-Visual-Studio-Code: 403 Client Error: Forbidden for url: https://support.enthought.com/hc/en-us/articles/360038600051-Canopy-GUI-end-of-life-transition-to-the-Enthought-Deployment-Manager-EDM-and-Visual-Studio-Code
```

Looks like crawling in this website is not allowed: link is ok in the browser, but fails with `curl` or sphinx's linkcheck. Added to ignored links.

```
library/readline.rst:20: [broken] https://tiswww.cwru.edu/php/chet/readline/rluserman.html#SEC9: Anchor 'SEC9' not found
```

The SEC9 anchor was about "Readline Init File" ([wayback machine link](https://web.archive.org/web/20210319041218/https://tiswww.cwru.edu/php/chet/readline/rluserman.html#SEC9)). I updated the anchor to match the same subject in the updated documentation.

```
faq/library.rst:780: [broken] https://twistedmatrix.com/trac/: 404 Client Error: Not Found for url: https://twisted.org/trac/
```

Updated URL to https://twisted.org/

```
whatsnew/2.5.rst:879: [broken] https://unix.org/version2/whatsnew/lp64_wp.html: HTTPSConnectionPool(host='unix.org', port=443): Max retries exceeded with url: /version2/whatsnew/lp64_wp.html (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self-signed certificate in certificate chain (_ssl.c:1129)')))
```

GoDaddy-hosted website, and GoDaddy's certificate chain is not installed causing curl and linkcheck to fail. Using a web browser works, though. So I added this to the ignored links list.

```
whatsnew/changelog.rst:22339: [broken] https://www.openssl.org/docs/man1.1.0/ssl/SSL_CTX_set_min_proto_version.html: 404 Client Error: Not Found for url: https://www.openssl.org/docs/man1.1.0/ssl/SSL_CTX_set_min_proto_version.html
```

Updated the URL, using the version 1.1.1 published as the closest possible to 1.1.0 mentioned in the paragraph.

```
library/zipfile.rst:10: [broken] https://github.com/python/cpython/tree/main/Lib/zipfile.py: 404 Client Error: Not Found for url: https://github.com/python/cpython/tree/main/Lib/zipfile.py
```

zipfile is a package since #98103. This change is post-3.11, hence a backport must not include this or will cause another 'broken' entry by linkcheck.

<!-- gh-issue-number: gh-103484 -->
* Issue: gh-103484
<!-- /gh-issue-number -->

rffontenelle committed:
Doc: Fix broken links reported by linkcheck

merwok commented on Doc/conf.py:
```diff
@@ -263,6 +263,26 @@
+
+linkcheck_ignore = [
+    # A code sample in Misc/NEWS.d/3.9.0a1.rst
+    'https:\/\/$',
```
Does linkcheck extract URLs from code blocks? or is that one badly marked-up?

rffontenelle replied:
It is not in a code-block.

![Screenshot_20230418-121724_Firefox.jpg](https://user-images.githubusercontent.com/1571783/232823458-b72ca683-5aea-41b8-bf1a-f64423a3741c.jpg)

rffontenelle replied:
I'm going to try code-blocking it and see if it solves without the ignore links entry

hugovk replied:
What happens if the code sample is wrapped in double backticks?

rffontenelle replied:
I'll run a test and will let you know.

merwok replied:
> It is not in a code-block.

Yep that’s the real bug!

rffontenelle replied:
Fixed. Made it literal (double backticks around the code sample) and linkcheck reported ok. Hence, I removed the ignore link entry.

merwok commented on Doc/faq/library.rst:
```diff
@@ -780,8 +780,8 @@ socket to :meth:`select.select` to check if it's writable.
-   The third-party `Twisted <https://twistedmatrix.com/trac/>`_ library is
-   a popular and feature-rich alternative.
+   The third-party `Twisted <https://twisted.org/>`_ library is a popular and
+   feature-rich alternative.
```
It’s not very important here, but minimizing diffs help history/blame queries.

```suggestion
   The third-party `Twisted <https://twisted.org/>`_ library is
   a popular and feature-rich alternative.
```

rffontenelle replied:
No problem. If the idea is to minimize diff, I'll squash & force-push your suggestion.

merwok replied:
Force pushes on github do not make a great experience for reviewers (comments get detached, notifications lead to page without any diff, etc), and are not useful here as Python PRs are always squash merged.  Normal commits are fine!

(See https://devguide.python.org/getting-started/pull-request-lifecycle/index.html for our little quirks.)

rffontenelle replied:
Ok, thanks for the feedback

hugovk commented on Doc/distributing/index.rst:
```diff
@@ -129,14 +129,10 @@ involved in creating and publishing a project:
-   https://packaging.python.org/tutorials/packaging-projects/#uploading-the-distribution-archives
-.. _The .pypirc file: \
-   https://packaging.python.org/specifications/pypirc/
+.. _Project structure: https://packaging.python.org/tutorials/packaging-projects/#packaging-python-projects
+.. _Building and packaging the project: https://packaging.python.org/tutorials/packaging-projects/#creating-the-package-files
+.. _Uploading the project to the Python Package Index: https://packaging.python.org/tutorials/packaging-projects/#uploading-the-distribution-archives
+.. _The .pypirc file: https://packaging.python.org/specifications/pypirc/
```
Should be fine just removing the ` \`? At least it renders okay.

```suggestion
.. _Project structure:
    https://packaging.python.org/tutorials/packaging-projects/#packaging-python-projects
.. _Building and packaging the project:
   https://packaging.python.org/tutorials/packaging-projects/#creating-the-package-files
.. _Uploading the project to the Python Package Index:
   https://packaging.python.org/tutorials/packaging-projects/#uploading-the-distribution-archives
.. _The .pypirc file:
   https://packaging.python.org/specifications/pypirc/
```

merwok replied:
That seems wrong, it looks like directives (which would need double colon) rather than link markup.
Best to have it on one line, that’s reST standard.

rffontenelle replied:
@hugovk Indeed, I tested and linkcheck doesn't complain if the `\` is removed.

How do we proceed here: One line or in a new line without `\`?

merwok replied:
One line please!

rffontenelle replied:
Ok then.

AA-Turner replied:
> Best to have it on one line, that’s reST standard.

reST specifies that whitespace is ignored specifically to allow breaking over multiple lines! E.g. when lines *must* be wrapped at a certain number of characters.

See https://www.docutils.org/docs/ref/rst/restructuredtext.html#hyperlinks, specifically the following section:

![image](https://user-images.githubusercontent.com/9087854/233499602-b95d6b78-445c-4f7c-a503-d3591bdf73a8.png)

A

merwok replied:
Ah that’s convenient for reST email for example!

Let me replace «that’s reST standard» with «that’s what’s commonly seen»

hugovk commented on Doc/library/stdtypes.rst:
```diff
@@ -1605,8 +1605,8 @@ expression support in the :mod:`re` module).
-   `described in section 3.13 of the Unicode Standard
-   <http://www.unicode.org/versions/Unicode15.0.0/ch03.pdf#G53253>`__.
+   `described in section 3.13 'Default Case Folding' of the Unicode Standard
+   <http://www.unicode.org/versions/Unicode15.0.0/ch03.pdf>`__.
```
It also redirects to https, let's include it here:
```suggestion
   <https://www.unicode.org/versions/Unicode15.0.0/ch03.pdf>`__.
```

hugovk commented on Misc/NEWS.d/3.7.0b2.rst:
```diff
@@ -357,7 +357,7 @@ Wirtel
 Add TLSVersion constants and SSLContext.maximum_version / minimum_version
 attributes. The new API wraps OpenSSL 1.1
-https://www.openssl.org/docs/man1.1.0/ssl/SSL_CTX_set_min_proto_version.html
+https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_set_min_proto_version.html
```
Maybe we should use an archive version here, as this text says it added TLSVersion constants, and 1.1.1 added a `TLS1_3_VERSION` not present in the 1.1.0 archive.

```suggestion
https://web.archive.org/web/20180309043602/https://www.openssl.org/docs/man1.1.0/ssl/SSL_CTX_set_min_proto_version.html
```

hugovk commented on Misc/NEWS.d/3.8.0a1.rst:
```diff
@@ -5951,7 +5951,7 @@ Wirtel
 Add TLSVersion constants and SSLContext.maximum_version / minimum_version
 attributes. The new API wraps OpenSSL 1.1
-https://www.openssl.org/docs/man1.1.0/ssl/SSL_CTX_set_min_proto_version.html
+https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_set_min_proto_version.html
```
Likewise
```suggestion
https://web.archive.org/web/20180309043602/https://www.openssl.org/docs/man1.1.0/ssl/SSL_CTX_set_min_proto_version.html
```

hugovk reviewed (commented):
Thanks for the PR! A few suggestions.

rffontenelle committed:
Doc: Make mark-up code as literal

rffontenelle committed:
Apply suggestions from code review

- Remove extra diff line in faq/library.rst (merwok)
- Use HTTPS to link Unicode 15.0.0 to solve a redirect (hugovk)
- Use wayback machine link for openssl 1.1.0 instead of linking 1.1.1, "as this text mentions a feature from 1.1.0" (hugovk)

Co-authored-by: Éric <merwok@netwok.org>
Co-authored-by: Hugo van Kemenade <hugovk@users.noreply.github.com>

hugovk commented on Doc/conf.py:
```diff
@@ -263,6 +263,24 @@
+    # The crawler gets "Anchor not found"
+    r'https://github.com.+?#.*',
+    r'https://devguide.python.org.+?/#.*',
+    r'https://developer.apple.com/documentation/.+?#.*',
```
Let's alphabetise this section, but makes it easier to diff other projects if they also have them:

```suggestion
    r'https://developer.apple.com/documentation/.+?#.*',
    r'https://devguide.python.org.+?/#.*',
    r'https://github.com.+?#.*',
```

rffontenelle committed:
Doc: Alphabetize items in linkcheck_ignore

Co-authored-by: Hugo van Kemenade <hugovk@users.noreply.github.com>

hugovk reviewed (approved):
Thank you!

AA-Turner commented on Doc/conf.py:
```diff
@@ -263,6 +263,24 @@
 }
 
+linkcheck_anchors_ignore = [
+    # ignore anchors that starts with a '/', e.g. Wikipedia media:
```
Tiny suggestion:
```suggestion
    # ignore anchors that start with a '/', e.g. Wikipedia media files:
```

AA-Turner commented on Doc/howto/functional.rst:
```diff
@@ -1208,8 +1208,8 @@ General
 **Structure and Interpretation of Computer Programs**, by Harold Abelson and
-Gerald Jay Sussman with Julie Sussman.  Full text at
-https://mitpress.mit.edu/sicp/.  In this classic textbook of computer science,
+Gerald Jay Sussman with Julie Sussman.  The book can be found at
+https://mitpress.mit.edu/sicp.  In this classic textbook of computer science,
```
The MIT Press website just says that the book is out of print -- I found [another resource](https://mitp-content-server.mit.edu/books/content/sectbyfn/books_pres_0/6515/sicp.zip/index.html) on the MIT domain, but I don't know how long-term the link will be.

rffontenelle replied:
There is also wayback machine (long term available) to get the point in history where the book was freely available. I didn't linked it before as I thought it wouldn't be appropriate, considering they are selling it now 🤷‍♂️

AA-Turner reviewed (approved):
Thank you!

A

rffontenelle committed:
Doc: Improve comment in sphinx conf

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>
